### PR TITLE
Add std::complex support to DataPostprocessor

### PR DESCRIPTION
**Pull request by dangars**
Hi, I would like to add support for a vector valued `std::complex` DataPostprocessor but I think that I took the wrong approach.

Could you give me a hint :) ?

I think that it should be feasible because the machinery to write a vector valued `std::complex` solution is already in place.

The members in `DataPostprocessor` are `double` and I have the feeling that `add_data_vector()` converts any `Vector<number>` to `Vector<double>`. That is a problem because `std::complex` can not be converted to double.

### CODE REVIEW

File: include/deal.II/numerics/data_postprocessor.h [hunk 1]
```diff
@@ -250,7 +250,7 @@ namespace DataPostprocessorInputs
      * or DataPostprocessorTensor may pass this flag to the constructor of
      * these three classes.
      */
-    std::vector<Tensor<1, spacedim>> solution_gradients;
+    std::vector<Tensor<1, spacedim, Number>> solution_gradients;
```
Comment (dangars) [inline include/deal.II/numerics/data_postprocessor.h]: This is `double` and I have the feeling that `DataPostprocessor` will convert any `number` to `double`. This wouldn't work for `std::complex`

File: include/deal.II/numerics/data_out_dof_data.h [hunk 2]
```diff
@@ -1094,12 +1113,17 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::add_data_vector(
...
 template <typename VectorType>
 void
-DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::add_data_vector(
-  const VectorType &                                        vec,
-  const DataPostprocessor<DoFHandlerType::space_dimension> &data_postprocessor)
+DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim, Number>::
+  add_data_vector(const VectorType &vec,
+                  const DataPostprocessor<DoFHandlerType::space_dimension,
+                                          typename VectorType::value_type>
+                    &data_postprocessor)
```
Comment (dangars) [inline include/deal.II/numerics/data_out_dof_data.h]: I made the type of `DataPostprocessor` equal to the type of the `Vector`. But I have the feeling that you do a conversion. For example `Vector` can be `float` and you convert it to `double`. This wouldn't work for `std::complex`

Why you convert `float` to `double`? It is possible to store `float` in the vtk/vtu file, there is no need for a conversion.

**Comment by bangerth** [comment]
`DataOut` and friends already support complex numbers: solution vectors are just separated into real and imaginary parts to create two scalar fields out of things. I implemented this as part of the (still unfinished) step-58. The two scalar parts will then also be passed as two solution components to the `DataPostprocessor`. So in some sense everything you want already works.

Why did I do it this way, rather than making everything templated on the scalar type? Because output formats generally do not understand anything other than real-valued types. So at some point one has to take complex-valued types apart and put them into two components. This currently happens just before we call the `DataPostprocessor` and/or lower the solution vectors to the `Patch` format.

In other words, I think that everything you need is already there. I'm not sure we need to make things more complicated beyond this :-)

**Comment by dangars** [comment]
Thanks! I understand now. It almost works :)

The current implementation of `DataPostprocessor` works for a `std::complex` scalar but I think that it fails for a `std::complex` vector (unless it is a bug in my program)

I close this PR and open the issue #8625 We can continue the conversation in the issue.
